File: OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseUnitMobileNotSupportedFragment.java
package org.edx.mobile.view;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;

import com.android.billingclient.api.BillingResult;
import com.android.billingclient.api.Purchase;

import org.edx.mobile.R;
import org.edx.mobile.databinding.FragmentCourseUnitGradeBinding;
import org.edx.mobile.inapppurchases.BillingProcessor;
import org.edx.mobile.model.api.AuthorizationDenialReason;
import org.edx.mobile.model.course.BlockType;
import org.edx.mobile.model.course.CourseComponent;
import org.edx.mobile.util.BrowserUtil;
import org.edx.mobile.util.UiUtils;

public class CourseUnitMobileNotSupportedFragment extends CourseUnitFragment {
    private FragmentCourseUnitGradeBinding binding;
    private BillingProcessor billingProcessor;

    public static CourseUnitMobileNotSupportedFragment newInstance(@NonNull CourseComponent unit, @NonNull boolean isSelfPaced, @NonNull String price) {
        final CourseUnitMobileNotSupportedFragment fragment = new CourseUnitMobileNotSupportedFragment();
        final Bundle args = new Bundle();
        args.putSerializable(Router.EXTRA_COURSE_UNIT, unit);
        args.putBoolean(Router.EXTRA_IS_SELF_PACED, isSelfPaced);
        args.putString(Router.EXTRA_PRICE, price);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_course_unit_grade, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        if (unit.getAuthorizationDenialReason() == AuthorizationDenialReason.FEATURE_BASED_ENROLLMENTS) {
            if (environment.getRemoteFeaturePrefs().isValuePropEnabled()) {
                boolean isSelfPaced = getBooleanArgument(Router.EXTRA_IS_SELF_PACED, false);
                String price = getStringArgument(Router.EXTRA_PRICE);
                binding.containerLayoutNotAvailable.setVisibility(View.GONE);
                binding.llGradedContentLayout.setVisibility(View.VISIBLE);
                setUpUpgradeButton(isSelfPaced, price);
                binding.toggleShow.setOnClickListener(v -> {
                    boolean showMore = binding.layoutUpgradeFeature.containerLayout.getVisibility() == View.GONE;
                    binding.layoutUpgradeFeature.containerLayout.setVisibility(showMore ? View.VISIBLE : View.GONE);
                    binding.toggleShow.setText(getText(showMore ? R.string.course_modal_graded_assignment_show_less : R.string.course_modal_graded_assignment_show_more));
                    environment.getAnalyticsRegistry().trackValuePropShowMoreLessClicked(unit.getCourseId(),
                            unit.getId(), price, isSelfPaced, showMore);
                });
            } else {
                binding.containerLayoutNotAvailable.setVisibility(View.VISIBLE);
                binding.llGradedContentLayout.setVisibility(View.GONE);
                binding.contentErrorIcon.setImageDrawable(UiUtils.INSTANCE.getDrawable(requireContext(), R.drawable.ic_lock));
                binding.notAvailableMessage.setText(R.string.not_available_on_mobile);
                binding.notAvailableMessage2.setVisibility(View.GONE);
            }
        } else {
            binding.containerLayoutNotAvailable.setVisibility(View.VISIBLE);
            binding.llGradedContentLayout.setVisibility(View.GONE);
            binding.contentErrorIcon.setImageDrawable(UiUtils.INSTANCE.getDrawable(requireContext(), R.drawable.ic_laptop));
            binding.notAvailableMessage.setText(unit.getType() == BlockType.VIDEO ?
                    R.string.video_only_on_web_short : R.string.assessment_not_available);
            binding.notAvailableMessage2.setVisibility(View.VISIBLE);
        }

        binding.viewOnWebButton.setOnClickListener(v -> {
            BrowserUtil.open(getActivity(), unit.getWebUrl(), true);
            environment.getAnalyticsRegistry().trackOpenInBrowser(unit.getId(), unit.getCourseId(),
                    unit.isMultiDevice(), unit.getBlockId());
        });
    }

    private void setUpUpgradeButton(boolean isSelfPaced, String price) {
        if (environment.getConfig().isIAPEnabled()) {
            binding.layoutUpgradeBtn.getRoot().setVisibility(View.VISIBLE);
            binding.layoutUpgradeBtn.btnUpgrade.setOnClickListener(v -> {
                enableUpgradeButton(false);
                purchaseProduct("org.edx.mobile.test_product");
                environment.getAnalyticsRegistry().trackUpgradeNowClicked(unit.getCourseId(),
                        price, unit.getId(), isSelfPaced);
            });

            billingProcessor = new BillingProcessor(requireContext(), new BillingProcessor.BillingFlowListeners() {
                @Override
                public void onPurchaseCancel() {
                    enableUpgradeButton(true);
                }

                @Override
                public void onBillingSetupFinished(@NonNull BillingResult billingResult) {
                    // Nothing do here
                }

                @Override
                public void onBillingServiceDisconnected() {
                    // Nothing do here
                }

                @Override
                public void onPurchaseComplete(@NonNull Purchase purchase) {
                    enableUpgradeButton(true);
                }
            });
        } else {
            binding.layoutUpgradeBtn.getRoot().setVisibility(View.GONE);
        }
    }

    private void enableUpgradeButton(boolean enable) {
        binding.layoutUpgradeBtn.btnUpgrade.setVisibility(enable ? View.VISIBLE : View.GONE);
        binding.layoutUpgradeBtn.loadingIndicator.setVisibility(!enable ? View.VISIBLE : View.GONE);
    }

    private void purchaseProduct(String productId) {
        if (billingProcessor != null) {
            billingProcessor.purchaseItem(requireActivity(), productId);
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        if (unit.getAuthorizationDenialReason() == AuthorizationDenialReason.FEATURE_BASED_ENROLLMENTS && environment.getRemoteFeaturePrefs().isValuePropEnabled()) {
            environment.getAnalyticsRegistry().trackLockedContentTapped(unit.getCourseId(), unit.getBlockId());
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if (billingProcessor != null) {
            billingProcessor.disconnect();
        }
    }
}

File: OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseUnitMobileNotSupportedFragment.kt
package org.edx.mobile.view

import android.content.DialogInterface
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.viewModels
import com.android.billingclient.api.BillingClient
import com.android.billingclient.api.ProductDetails
import dagger.hilt.android.AndroidEntryPoint
import org.edx.mobile.R
import org.edx.mobile.databinding.FragmentCourseUnitGradeBinding
import org.edx.mobile.event.IAPFlowEvent
import org.edx.mobile.exception.ErrorMessage
import org.edx.mobile.extenstion.isNotVisible
import org.edx.mobile.extenstion.setImageDrawable
import org.edx.mobile.extenstion.setVisibility
import org.edx.mobile.http.HttpStatus
import org.edx.mobile.inapppurchases.getPriceAmount
import org.edx.mobile.model.api.AuthorizationDenialReason
import org.edx.mobile.model.api.EnrolledCoursesResponse
import org.edx.mobile.model.course.CourseComponent
import org.edx.mobile.model.iap.IAPFlowData
import org.edx.mobile.model.iap.IAPFlowData.IAPAction
import org.edx.mobile.module.analytics.Analytics
import org.edx.mobile.module.analytics.Analytics.Events
import org.edx.mobile.module.analytics.Analytics.Screens
import org.edx.mobile.module.analytics.InAppPurchasesAnalytics
import org.edx.mobile.util.AppConstants
import org.edx.mobile.util.BrowserUtil
import org.edx.mobile.util.InAppPurchasesException
import org.edx.mobile.util.ResourceUtil
import org.edx.mobile.util.observer.EventObserver
import org.edx.mobile.viewModel.InAppPurchasesViewModel
import org.edx.mobile.wrapper.InAppPurchasesDialog
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import javax.inject.Inject

@AndroidEntryPoint
class CourseUnitMobileNotSupportedFragment : CourseUnitFragment() {

    private lateinit var binding: FragmentCourseUnitGradeBinding

    private val iapViewModel: InAppPurchasesViewModel by viewModels()

    @Inject
    lateinit var iapAnalytics: InAppPurchasesAnalytics

    @Inject
    lateinit var iapDialog: InAppPurchasesDialog

    private var price: String? = null

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentCourseUnitGradeBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val isUpgradeable = getBooleanArgument(Router.EXTRA_IS_UPGRADEABLE, true)
        if (AuthorizationDenialReason.FEATURE_BASED_ENROLLMENTS == unit?.authorizationDenialReason) {
            if (environment.featuresPrefs.isValuePropEnabled && isUpgradeable) {
                showGradedContent()
            } else {
                showNotAvailableOnMobile(isLockedContent = true)
            }
        } else {
            showNotAvailableOnMobile(isLockedContent = false)
        }
    }

    private fun showGradedContent() {
        unit?.let { unit ->
            val isSelfPaced = getBooleanArgument(Router.EXTRA_IS_SELF_PACED, false)
            val isPurchaseEnabled = unit.courseSku.isNullOrEmpty().not() &&
                    environment.featuresPrefs.isIAPEnabledForUser(environment.loginPrefs.isOddUserId)

            binding.containerLayoutNotAvailable.root.setVisibility(false)
            binding.llGradedContentLayout.setVisibility(true)
            var experimentGroup: String? = null
            if (environment.featuresPrefs.isIAPExperimentEnabled) {
                experimentGroup = if (environment.loginPrefs.isOddUserId) {
                    Analytics.Values.TREATMENT
                } else {
                    Analytics.Values.CONTROL
                }
            }
            environment.analyticsRegistry.trackLockedContentTapped(unit.courseId, unit.blockId)
            environment.analyticsRegistry.trackValuePropMessageViewed(
                unit.courseId,
                Screens.COURSE_UNIT,
                (unit.courseSku.isNullOrEmpty().not() && environment.featuresPrefs.isIAPEnabled),
                experimentGroup,
                unit.id
            )
            if (isPurchaseEnabled) {
                iapAnalytics.initCourseValues(
                    courseId = unit.courseId,
                    isSelfPaced = isSelfPaced,
                    flowType = IAPFlowData.IAPFlowType.USER_INITIATED.value(),
                    screenName = Screens.COURSE_COMPONENT,
                    componentId = unit.id
                )
                initIAPObserver()
                // Shimmer container taking sometime to get ready and perform the animation, so
                // by adding the some delay fixed that issue for lower-end devices, and for the
                // proper animation.
                binding.layoutUpgradeBtn.shimmerViewContainer.postDelayed({
                    iapViewModel.initializeProductPrice(unit.courseSku)
                }, 1500)
                binding.layoutUpgradeBtn.btnUpgrade.isEnabled = false
            } else {
                binding.layoutUpgradeBtn.root.setVisibility(false)
            }

            binding.toggleShow.setOnClickListener {
                val showMore = binding.layoutUpgradeFeature.containerLayout.isNotVisible()
                binding.layoutUpgradeFeature.containerLayout.setVisibility(showMore)
                binding.toggleShow.text = getText(
                    if (showMore) {
                        R.string.course_modal_graded_assignment_show_less
                    } else {
                        R.string.course_modal_graded_assignment_show_more
                    }
                )
                environment.analyticsRegistry.trackValuePropShowMoreLessClicked(
                    unit.courseId, unit.id, price, isSelfPaced, showMore
                )
            }
        }
    }

    private fun showNotAvailableOnMobile(isLockedContent: Boolean) {
        binding.llGradedContentLayout.setVisibility(false)
        binding.containerLayoutNotAvailable.apply {
            root.setVisibility(true)
            notAvailableMessage2.setVisibility(!isLockedContent)

            if (isLockedContent) {
                contentErrorIcon.setImageDrawable(R.drawable.ic_lock)
                notAvailableMessage.setText(R.string.not_available_on_mobile)
            } else {
                contentErrorIcon.setImageDrawable(R.drawable.ic_laptop)
                notAvailableMessage.setText(
                    if (unit?.isVideoBlock == true) R.string.video_only_on_web_short
                    else R.string.assessment_not_available
                )
            }

            viewOnWebButton.setOnClickListener {
                unit?.let {
                    BrowserUtil.open(activity, it.webUrl, true)
                    environment.analyticsRegistry.trackOpenInBrowser(
                        it.id, it.courseId, it.isMultiDevice, it.blockId
                    )
                }
            }
        }
    }

    private fun initIAPObserver() {
        initializeBaseObserver()

        iapViewModel.productPrice.observe(viewLifecycleOwner, EventObserver {
            setUpUpgradeButton(it)
        })

        iapViewModel.launchPurchaseFlow.observe(viewLifecycleOwner, EventObserver {
            iapViewModel.purchaseItem(
                requireActivity(),
                environment.loginPrefs.userId,
                unit?.courseSku
            )
        })

        iapViewModel.productPurchased.observe(viewLifecycleOwner, EventObserver {
            EventBus.getDefault().post(IAPFlowEvent(IAPAction.SHOW_FULL_SCREEN_LOADER, it))
        })

        iapViewModel.showLoader.observe(viewLifecycleOwner, EventObserver {
            enableUpgradeButton(!it)
        })

        iapViewModel.errorMessage.observe(viewLifecycleOwner, EventObserver { errorMessage ->
            handleIAPException(errorMessage)
        })
    }

    private fun handleIAPException(errorMessage: ErrorMessage) {
        var retryListener: DialogInterface.OnClickListener? = null
        if (HttpStatus.NOT_ACCEPTABLE == (errorMessage.throwable as InAppPurchasesException).httpErrorCode) {
            retryListener = DialogInterface.OnClickListener { _, _ ->
                EventBus.getDefault().post(
                    IAPFlowEvent(IAPAction.SHOW_FULL_SCREEN_LOADER, iapViewModel.iapFlowData)
                )
            }
        } else if (BillingClient.BillingResponseCode.USER_CANCELED == errorMessage.getHttpErrorCode()) {
            iapAnalytics.trackIAPEvent(eventName = Events.IAP_PAYMENT_CANCELED)
            return
        } else if (errorMessage.canRetry()) {
            retryListener = DialogInterface.OnClickListener { _, _ ->
                when (errorMessage.requestType) {
                    ErrorMessage.PRICE_CODE -> {
                        iapViewModel.initializeProductPrice(unit?.courseSku)
                    }
                }
            }
        }
        iapDialog.handleIAPException(
            fragment = this@CourseUnitMobileNotSupportedFragment,
            errorMessage = errorMessage,
            retryListener = retryListener
        )
    }

    private fun setUpUpgradeButton(productDetails: ProductDetails.OneTimePurchaseOfferDetails) {
        price = productDetails.formattedPrice
        binding.layoutUpgradeBtn.root.setVisibility(true)

        binding.layoutUpgradeBtn.btnUpgrade.text =
            ResourceUtil.getFormattedString(
                resources,
                R.string.label_upgrade_course_button,
                AppConstants.PRICE,
                price
            ).toString()
        // The app get the product details instantly, so add some wait to perform
        // animation at least one cycle.
        binding.layoutUpgradeBtn.shimmerViewContainer.postDelayed({
            binding.layoutUpgradeBtn.shimmerViewContainer.hideShimmer()
            binding.layoutUpgradeBtn.btnUpgrade.isEnabled = true
        }, 500)

        binding.layoutUpgradeBtn.btnUpgrade.setOnClickListener {
            iapAnalytics.trackIAPEvent(Events.IAP_UPGRADE_NOW_CLICKED)
            unit?.courseSku?.let { productId ->
                iapViewModel.startPurchaseFlow(
                    productId,
                    productDetails.getPriceAmount(),
                    productDetails.priceCurrencyCode,
                )
            } ?: iapDialog.showPreUpgradeErrorDialog(this)
        }
    }

    private fun enableUpgradeButton(enable: Boolean) {
        binding.layoutUpgradeBtn.btnUpgrade.setVisibility(enable)
        binding.layoutUpgradeBtn.loadingIndicator.setVisibility(!enable)
    }

    override fun onResume() {
        super.onResume()
        EventBus.getDefault().register(this)
    }

    override fun onPause() {
        EventBus.getDefault().unregister(this)
        super.onPause()
    }

    @Subscribe
    fun onEventMainThread(event: IAPFlowEvent) {
        if (isResumed && event.flowAction == IAPAction.PURCHASE_FLOW_COMPLETE) {
            unit?.let { updateCourseUnit(it.courseId, it.id) }
        }
    }

    companion object {
        @JvmStatic
        fun newInstance(
            unit: CourseComponent,
            courseData: EnrolledCoursesResponse,
        ): CourseUnitMobileNotSupportedFragment {
            val fragment = CourseUnitMobileNotSupportedFragment()
            val args = Bundle()
            args.putSerializable(Router.EXTRA_COURSE_UNIT, unit)
            args.putBoolean(Router.EXTRA_IS_SELF_PACED, courseData.course.isSelfPaced)
            args.putBoolean(Router.EXTRA_IS_UPGRADEABLE, courseData.isUpgradeable)
            fragment.arguments = args
            return fragment
        }
    }
}
